idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt

```diff
@@ -61,6 +66,10 @@ internal class KtFirExpressionTypeProvider(
         getExpectedTypeByReturnExpression(expression)
             ?: getExpressionTypeByIfOrBooleanCondition(expression)
             ?: getExpectedTypeOfFunctionParameter(expression)
+            ?: getExpectedTypeOfInfixFunctionParameter(expression)
+            ?: getExpectedTypeByVariableAssignment(expression)
+            ?: getExpectedTypeByPropertyDeclaration(expression)
+            ?: getExpectedTypeByFunctionExpressionBody(expression)
```

Comment:
Maybe this is written this way before, but I'm not sure about this...

For example, say, in `getExpectedTypeOfInfixFunctionParameter`, `unwrapQualified` succeeds, but some other fails, e.g., unresolved return type. Then, all the following calls are simply redundant, this chain will keep checking expression type. Maybe not a big deal, but somehow simple type-based `when` could be helpful.

Reply:
That's a good idea! It will require some significant rework though, to support moving forward to the next "matcher" or stop completely. I am not sure how a type-based `when` can work, given that we have to unwrap _and_ do a basic check (lambda to `unwrapQualified`) before proceeding.

Reply:
To be clear, it's a good idea, but I am not planning to implement that in this PR.